src/NuGet.Core/NuGet.Protocol/Resources/VulnerabilityInfoResourceV3.cs

```diff
@@ -0,0 +1,299 @@
+                throw new FatalProtocolException(message);
+            }
+
+            return data;
+        }
+
+        /// <inheritdoc cref="IVulnerabilityInfoResource.GetVulnerabilityInfoAsync(SourceCacheContext, ILogger, CancellationToken)"/>
+        public async Task<GetVulnerabilityInfoResult> GetVulnerabilityInfoAsync(SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
+        {
+            List<Exception>? exceptions = null;
+            List<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>? knownVulnerabilities = null;
+
+            try
+            {
+                IReadOnlyList<V3VulnerabilityIndexEntry> indexEntries = await GetVulnerabilityFilesAsync(cacheContext, logger, cancellationToken);
+
+                if (indexEntries.Count == 0)
+                {
+                    knownVulnerabilities = new();
+                    return new GetVulnerabilityInfoResult(knownVulnerabilities, ToAggregateException(exceptions));
+                }
+
+                var tasks = new Task<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>>[indexEntries.Count];
+                for (int i = 0; i < tasks.Length; i++)
+                {
+                    V3VulnerabilityIndexEntry indexEntry = indexEntries[i];
+                    tasks[i] = GetVulnerabilityDataAsync(indexEntry, cacheContext, logger, cancellationToken);
+                }
+
+                indexEntries = GetValidIndexEntries(indexEntries, ref exceptions);
+
+                await Task.WhenAll(tasks);
+
+                for (int i = 0; i < tasks.Length; i++)
+                {
+                    try
+                    {
+                        IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>? data = await tasks[i];
+                        if (data != null)
+                        {
+                            if (knownVulnerabilities == null)
+                            {
+                                knownVulnerabilities = new(indexEntries.Count);
+                            }
+                            knownVulnerabilities.Add(data);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        AddException(ref exceptions, ex);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                AddException(ref exceptions, ex);
+            }
+
+            GetVulnerabilityInfoResult result = new(knownVulnerabilities, ToAggregateException(exceptions));
+            return result;
+
+            static void AddException(ref List<Exception>? exceptions, Exception ex)
+            {
+                if (exceptions == null)
+                {
+                    exceptions = new();
+                }
+                exceptions.Add(ex);
+            }
```

Comment:
There are two definitions of AddException with different order of parameters. Consider making this just one static method definition on the class.